src/server/scripts/Zandalar/Underrot/boss_cragmaw_the_infested.cpp

```diff
@@ -0,0 +1,377 @@
+{
+    boss_cragmaw_the_infested(Creature* creature) : BossAI(creature, DATA_CRAGMAW_THE_INFESTED) { }
+
+    void JustDied(Unit* /*killer*/) override
+    {
+        _JustDied();
+        instance->SendEncounterUnit(ENCOUNTER_FRAME_DISENGAGE, me);
+    }
+
+    void JustAppeared() override
+    {
+        if (instance->GetData(DATA_CRAGMAW_CRAWG_EATING))
+            return;
+
+        DoCastSelf(SPELL_CRAWG_EATING);
+        me->SetEmoteState(EMOTE_STATE_EAT);
+        if (TempSummon* summon = me->SummonCreature(NPC_FETID_MAGGOT, FetidMaggotSpawn))
+        {
+            _fetidMaggotGuid = summon->GetGUID();
+            summon->CastSpell(nullptr, SPELL_FEIGN_DEATH, true);
+            summon->SetUnitFlag(UnitFlags(UNIT_FLAG_IMMUNE_TO_PC | UNIT_FLAG_IMMUNE_TO_NPC | UNIT_FLAG_UNINTERACTIBLE));
+        }
+    }
+
+    void EnterEvadeMode(EvadeReason /*why*/) override
+    {
+        instance->SendEncounterUnit(ENCOUNTER_FRAME_DISENGAGE, me);
+        summons.DespawnAll();
+        _EnterEvadeMode();
+        _DespawnAtEvade();
+    }
+
+    void ScheduleSpells()
+    {
+        events.Reset();
+
+        if (GetDifficulty() == DIFFICULTY_NORMAL)
+            events.ScheduleEvent(EVENT_NORMAL_REQUEUE, 45s);
+
+        if (urand(0, 1) == 0)
+        {
+            events.ScheduleEvent(EVENT_CHARGE_SELECTOR, 9s);
+            events.ScheduleEvent(EVENT_INDIGESTION, 20s);
+            events.ScheduleEvent(EVENT_CHARGE_SELECTOR, 33s);
+        }
+        else
+        {
+            events.ScheduleEvent(EVENT_INDIGESTION, 9s);
+            events.ScheduleEvent(EVENT_CHARGE_SELECTOR, 21s);
+            events.ScheduleEvent(EVENT_CHARGE_SELECTOR, 41s);
+        }
+    }
+
+    void JustEngagedWith(Unit* who) override
+    {
+        BossAI::JustEngagedWith(who);
+        instance->SendEncounterUnit(ENCOUNTER_FRAME_ENGAGE, me, 1);
+        instance->SetData(DATA_CRAGMAW_CRAWG_EATING, 1);
+        ScheduleSpells();
+
+        if (Creature* fetidMaggot = ObjectAccessor::GetCreature(*me, _fetidMaggotGuid))
+            fetidMaggot->DespawnOrUnsummon();
+
+        if (IsHeroic() || GetDifficulty() == DIFFICULTY_MYTHIC || GetDifficulty() == DIFFICULTY_MYTHIC_KEYSTONE)
+        {
+            me->SetPowerType(POWER_ENERGY);
+            DoCast(SPELL_TANTRUM_ENERGY_BAR);
+            DoCast(SPELL_TANTRUM_PERIODIC_ENERGY);
+        }
+    }
+
+    void MovementInform(uint32 /*type*/, uint32 id) override
+    {
+        if (id == POINT_TANTRUM_START_RND_MOVEMENT)
+            me->GetMotionMaster()->MoveRandom(20.0f);
+    }
+
+    void OnChannelFinished(SpellInfo const* spell) override
+    {
+        if (spell->Id == SPELL_TANTRUM_INITIAL)
+            me->SetReactState(REACT_AGGRESSIVE);
+    };
+
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
+
+        events.Update(diff);
+
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
+
+        if (me->GetPower(POWER_ENERGY) >= 100)
```

Comment:
I'm not sure if we want to check on every single tick whether the energy is filled. Isn't it best to use EventMap for this? On retail, even if a creature reaches its required energy to do something, there's some sort of delay as if the check had a timer.